File: backend/api/services/collection.py
"""Collection coverage, sweeps and the quality caveats that go with them.

This is the "honest coverage" surface. It exists so that the gap between what we
meant to collect and what we actually collected is visible in the product rather
than discoverable by an examiner.
"""
import datetime as _dt
import json
import pathlib

ROOT = pathlib.Path(__file__).resolve().parents[2]
BASKET = ROOT / "apix" / "data" / "route_basket.json"

# The scheduled collection slot, matching `.github/workflows/collect.yml`
# (cron "15 5 * * *" = 05:15 UTC = 10:45 IST).  Every day is LABELLED with this
# nominal time, the way a statistics office labels a price to a reference
# period rather than to the minute an enumerator happened to write it down.
#
# It never overwrites `collected_at`.  The real timestamp stays in the database
# and in the API, because it has to keep agreeing with the git commit times and
# the public Actions logs -- our whole provenance argument rests on that, and a
# uniform stored timestamp would contradict a public record in the same repo.
NOMINAL_IST = "10:45"
_IST = _dt.timedelta(hours=5, minutes=30)


def _parse(ts: str) -> _dt.datetime:
    """`collected_at` / `started_at` are stored naive UTC."""
    return _dt.datetime.fromisoformat(ts.replace("T", " ").split(".")[0])


def _ist(ts: str) -> str:
    """UTC timestamp -> HH:MM in IST, for display beside the nominal slot."""
    return (_parse(ts) + _IST).strftime("%H:%M")


def _drift_minutes(ts: str) -> int:
    """Signed minutes between the nominal slot and when we actually collected."""
    t = _parse(ts) + _IST
    h, m = (int(x) for x in NOMINAL_IST.split(":"))
    return round((t - t.replace(hour=h, minute=m, second=0,
                                microsecond=0)).total_seconds() / 60)


def coverage(con, index_res) -> dict:
    """
    Backfilled rows (source LIKE 'external\\_%') are excluded from every table
    and count below. This page's whole point is documenting what *we*
    collected and how -- run timing, ethics, rate limits. A row imported from
    another team's public repo/API to plug the pre-window date gap has none of
    that: no `collection_run`, no real `run_id`, no rate-limit story. Counting
    it here would show a "collection day" with zero real attempts behind it,
    which is exactly the kind of thing this page exists to make visible, not
    produce. See `backfill` below for the honest, separately-labelled count.
    """
    basket = json.loads(BASKET.read_text())
    in_basket = [f"{r['origin']}-{r['destination']}" for r in basket["routes"]]

    LIVE = "source NOT LIKE 'external\\_%' ESCAPE '\\'"

    attempted = {f"{o}-{d}" for o, d in con.execute(
        "SELECT DISTINCT origin, destination FROM collection_run")}
    with_fares = {f"{o}-{d}" for o, d in con.execute(
        f"SELECT DISTINCT origin, destination FROM fare_observation WHERE {LIVE}")}

    backfill_rows = con.execute(
        "SELECT COUNT(*), MIN(substr(collected_at,1,10)), "
        "MAX(substr(collected_at,1,10)) FROM fare_observation "
        "WHERE source LIKE 'external\\_%' ESCAPE '\\'").fetchone()
    backfill_sources = [r[0] for r in con.execute(
        "SELECT DISTINCT source FROM fare_observation "
        "WHERE source LIKE 'external\\_%' ESCAPE '\\'")]
    backfill_routes = sorted({f"{o}-{d}" for o, d in con.execute(
        "SELECT DISTINCT origin, destination FROM fare_observation "
        "WHERE source LIKE 'external\\_%' ESCAPE '\\'")})

    days = []
    for date, n_obs in con.execute(
            f"SELECT substr(collected_at,1,10) d, COUNT(*) FROM fare_observation "
            f"WHERE {LIVE} GROUP BY d ORDER BY d"):
        runs = [r[0] for r in con.execute(
            f"SELECT DISTINCT run_id FROM fare_observation "
            f"WHERE substr(collected_at,1,10)=? AND {LIVE}", (date,))]
        attempts = con.execute(
            "SELECT COUNT(*), SUM(status='OK'), SUM(status!='OK') FROM collection_run "
            "WHERE substr(started_at,1,10)=?", (date,)).fetchone()
        lo, hi = con.execute(
            f"SELECT MIN(collected_at), MAX(collected_at) FROM fare_observation "
            f"WHERE substr(collected_at,1,10)=? AND {LIVE}", (date,)).fetchone()
        days.append({
            "date": date, "n_observations": n_obs,
            "n_cells": index_res["collection"]["cells_per_day"].get(date, 0),
            "n_runs": len(runs), "runs": sorted(runs),
            "attempts": attempts[0], "ok": attempts[1] or 0,
            "failed": attempts[2] or 0,
            # The scheduled slot every day is labelled against. Uniform by
            # construction -- it is the timetable, not a measurement.
            "nominal_time_ist": NOMINAL_IST,
            # What the clock actually read. Never overwritten: this is the
            # provenance trail, and it must keep matching the git commit times.
            "actual_ist": {"first": _ist(lo), "last": _ist(hi)},
            "drift_minutes": _drift_minutes(lo),
            "hour_spread": {"first": lo[11:16], "last": hi[11:16]},   # raw UTC
        })

    varies = len({d["actual_ist"]["first"][:2] for d in days}) > 1

    pax_total = sum(r["pax_cy"] for r in basket["routes"])
    pax_covered = sum(r["pax_cy"] for r in basket["routes"]
                      if f"{r['origin']}-{r['destination']}" in with_fares)

    return {
        "summary": {
            "observations": con.execute(
                f"SELECT COUNT(*) FROM fare_observation WHERE {LIVE}").fetchone()[0],
            "rows_considered": index_res["collection"]["rows_considered"],
            "rows_selected": index_res["collection"]["rows_selected"],
            "collection_runs": con.execute(
                "SELECT COUNT(*) FROM collection_run").fetchone()[0],
            "sweeps": con.execute(
                "SELECT COUNT(DISTINCT run_id) FROM collection_run").fetchone()[0],
            "days": len(days),
            "sources": [r[0] for r in con.execute(
                f"SELECT DISTINCT source FROM fare_observation WHERE {LIVE}")],
        },
        "backfill": {
            "rows": backfill_rows[0],
            "date_range": [backfill_rows[1], backfill_rows[2]] if backfill_rows[0] else None,
            "sources": backfill_sources,
            "routes": backfill_routes,
            "note": "Pre-window historical rows imported from other teams' public "
                    "repos/APIs to cover dates before our own collection started "
                    "(2026-09-10). Not a live sweep, no collection_run behind it, "
                    "and excluded from the index (config.INDEX_SOURCE keeps the "
                    "series on one price source). Kept separate from every count "
                    "and table above so this page still describes only what we "
                    "actually ran.",
        },
        "days": days,
        "clock": {
            "nominal_time_ist": NOMINAL_IST,
            "schedule": "cron 15 5 * * * (05:15 UTC / 10:45 IST), "
                        ".github/workflows/collect.yml",
            "nominal_note": "Every day is labelled with the scheduled slot, the "
                            "way a price is labelled to a reference period rather "
                            "than to the minute it was written down. The actual "
                            "clock time is kept beside it and is never overwritten.",
            "actual_ist_observed": [d["actual_ist"]["first"] for d in days],
            "max_abs_drift_minutes": max((abs(d["drift_minutes"]) for d in days),
                                         default=0),
            "varies_across_days": varies,
            "warning": "Actual collection hour varies across days, so part of the "
                       "measured movement is the clock, not the market. These were "
                       "the manual bootstrap days; the schedule above is what "
                       "unattended runs follow."
            if varies else None,
        },
        "routes": {
            "in_basket": len(in_basket),
            "attempted": len(attempted & set(in_basket)),
            "with_fares": len(with_fares & set(in_basket)),
            # A route the collector never tried is not a failed route. Keeping the
            # two apart matters: one is a coverage gap we chose, the other is a bug.
            "never_attempted": sorted(r for r in in_basket if r not in attempted),
            "attempted_but_no_fares": sorted(
                r for r in in_basket if r in attempted and r not in with_fares),
            "basket_pax_covered_pct": round(100 * pax_covered / pax_total, 2),
            "national_pax_covered_pct": round(
                100 * pax_covered / basket["national_total_pax"], 2),
        },
        "sweep_selection": {
            "rule": "Per (date, route, lead), the latest genuine sweep that covered "
                    "it wins. Selection is per cell, not per day: 2026-09-10 is the "
                    "union of four partial sweeps and has no single complete one.",
            "excluded_runs": index_res["collection"]["excluded_runs"],
            "genuineness_screen": "A sweep whose fetches arrive faster than the "
                                  "collector's own rate limit cannot be live "
                                  "collection. Nothing is excluded at present.",
        },
    }


def sweeps(con, limit: int = 50) -> dict:
    """Recent sweeps, rolled up -- the grain a activity feed wants.

    Every run carries `kind`, because not every run in this table feeds the
    index. The daily fare sweep does; the base-fare/tax split study does not --
    it prices a different concept on a different platform and is pinned out of
    the series by config.INDEX_SOURCE. Showing them in one undifferentiated
    list reads as "we collected three times today", which is not what happened.
    """
    from apix.index.config import INDEX_SOURCE
    rows = con.execute(
        """SELECT run_id, MIN(started_at), MAX(started_at), COUNT(*),
                  SUM(n_quotes), SUM(status='OK'), SUM(status!='OK'), source
           FROM collection_run GROUP BY run_id ORDER BY MIN(started_at) DESC
           LIMIT ?""", (limit,)).fetchall()
    out = []
    for rid, t0, t1, fetches, quotes, ok, bad, source in rows:
        import datetime
        span = (datetime.datetime.fromisoformat(t1)
                - datetime.datetime.fromisoformat(t0)).total_seconds()
        out.append({
            "run_id": rid, "date": t0[:10],
            "started": t0[11:19], "ended": t1[11:19],
            "fetches": fetches, "quotes": quotes or 0,
            "ok": ok or 0, "failed": bad or 0,
            "span_seconds": round(span, 1),
            "seconds_per_fetch": round(span / max(fetches - 1, 1), 2),
            "source": source,
            "feeds_index": source == INDEX_SOURCE,
            "kind": "Daily fare sweep" if source == INDEX_SOURCE
                    else "Base fare / tax study",
        })
    return {"n_sweeps": len(out), "sweeps": out}


def run_log(con, limit: int = 200) -> dict:
    rows = con.execute(
        """SELECT run_id, origin, destination, lead_time_days, departure_date,
                  source, status, n_quotes, elapsed_ms, error, started_at
           FROM collection_run ORDER BY started_at DESC LIMIT ?""", (limit,)).fetchall()
    return {"n_runs": len(rows), "runs": [
        {"run_id": r[0], "route": f"{r[1]}-{r[2]}", "lead_time_days": r[3],
         "departure_date": r[4], "source": r[5], "status": r[6],
         "n_quotes": r[7], "elapsed_ms": r[8], "error": r[9],
         "started_at": r[10]} for r in rows]}

File: src/server/services/collection.ts
import { sql, num, isoDate } from '../db';
import type { Vintage } from '../envelope';

/**
 * The collection log: what was actually swept, when, and how.
 *
 * The sweep-timing arithmetic is done in SQL with named timezones, not ported
 * to JavaScript. The Python original parses a naive-UTC string, adds a
 * hardcoded +5:30, then anchors against the same day's 10:45 -- a `Date`-based
 * port would silently apply the container's timezone (UTC on Vercel, IST on a
 * developer laptop) and produce plausible-looking wrong drift minutes. Naming
 * 'Asia/Kolkata' in the query removes the ambiguity entirely.
 */

type Row = Record<string, unknown>;

/** The nominal scheduled slot every day is labelled against. */
const NOMINAL_IST = '10:45';
const INDEX_SOURCE_NOTE = 'Daily fare sweep';

export async function coverage(v: Vintage) {
  const [
    basket, attemptedRows, withFaresRows, backfill, backfillSrc, backfillRoutes,
    dayRows, summary, cellRows, exclusions, runHeader,
  ] = await Promise.all([
    sql`SELECT origin, destination, pax_cy FROM route_basket ORDER BY origin, destination`,
    sql`SELECT DISTINCT origin, destination FROM collection_run`,
    sql`SELECT DISTINCT origin, destination FROM fare_observation
         WHERE NOT starts_with(source, 'external_')`,
    sql`SELECT COUNT(*)::int AS n, MIN(obs_date) AS lo, MAX(obs_date) AS hi
          FROM fare_observation WHERE starts_with(source, 'external_')`,
    sql`SELECT DISTINCT source FROM fare_observation
         WHERE starts_with(source, 'external_')`,
    sql`SELECT DISTINCT origin, destination FROM fare_observation
         WHERE starts_with(source, 'external_') ORDER BY origin, destination`,
    sql`
      WITH obs AS (
        SELECT obs_date, COUNT(*)::int AS n_obs,
               MIN(collected_at) AS lo, MAX(collected_at) AS hi,
               array_agg(DISTINCT run_id) AS runs
          FROM fare_observation
         WHERE NOT starts_with(source, 'external_')
         GROUP BY obs_date
      ), att AS (
        SELECT start_date,
               COUNT(*)::int AS attempts,
               COUNT(*) FILTER (WHERE status = 'OK')::int AS ok,
               COUNT(*) FILTER (WHERE status <> 'OK')::int AS failed
          FROM collection_run GROUP BY start_date
      )
      SELECT o.obs_date, o.n_obs, o.runs,
             COALESCE(a.attempts, 0) AS attempts,
             COALESCE(a.ok, 0) AS ok,
             COALESCE(a.failed, 0) AS failed,
             to_char(o.lo AT TIME ZONE 'Asia/Kolkata', 'HH24:MI') AS ist_first,
             to_char(o.hi AT TIME ZONE 'Asia/Kolkata', 'HH24:MI') AS ist_last,
             to_char(o.lo AT TIME ZONE 'UTC', 'HH24:MI') AS utc_first,
             to_char(o.hi AT TIME ZONE 'UTC', 'HH24:MI') AS utc_last,
             round(EXTRACT(EPOCH FROM (
                 (o.lo AT TIME ZONE 'Asia/Kolkata')
               - (date_trunc('day', o.lo AT TIME ZONE 'Asia/Kolkata')
                  + interval '10 hours 45 minutes')
             )) / 60)::int AS drift_minutes
        FROM obs o LEFT JOIN att a ON a.start_date = o.obs_date
       ORDER BY o.obs_date
    `,
    sql`
      SELECT (SELECT COUNT(*)::int FROM fare_observation
               WHERE NOT starts_with(source, 'external_')) AS observations,
             (SELECT COUNT(*)::int FROM collection_run) AS collection_runs,
             (SELECT COUNT(DISTINCT run_id)::int FROM collection_run) AS sweeps
    `,
    sql`SELECT obs_date, n_cells FROM collection_day_stat WHERE index_run_id = ${v.runId}`,
    sql`SELECT run_id, reason FROM sweep_exclusion WHERE index_run_id = ${v.runId}`,
    sql`SELECT n_obs_in, n_obs_selected FROM index_run WHERE id = ${v.runId}`,
  ]);

  // First-appearance order, not alphabetical: SQLite returned these in table
  // scan order and the published snapshot records that order.
  const sources = (await sql`
    SELECT source FROM fare_observation
     WHERE NOT starts_with(source, 'external_')
     GROUP BY source ORDER BY MIN(id)
  `) as Row[];

  const inBasket = (basket as Row[]).map((r) => `${r.origin}-${r.destination}`);
  const attempted = new Set(
    (attemptedRows as Row[]).map((r) => `${r.origin}-${r.destination}`),
  );
  const withFares = new Set(
    (withFaresRows as Row[]).map((r) => `${r.origin}-${r.destination}`),
  );
  const cellsBy = new Map(
    (cellRows as Row[]).map((r) => [isoDate(r.obs_date), num(r.n_cells)]),
  );

  const days = (dayRows as Row[]).map((d) => ({
    date: isoDate(d.obs_date),
    n_observations: num(d.n_obs),
    n_cells: cellsBy.get(isoDate(d.obs_date)) ?? 0,
    n_runs: (d.runs as string[]).length,
    runs: [...(d.runs as string[])].sort(),
    attempts: num(d.attempts),
    ok: num(d.ok),
    failed: num(d.failed),
    nominal_time_ist: NOMINAL_IST,
    actual_ist: { first: d.ist_first, last: d.ist_last },
    drift_minutes: num(d.drift_minutes),
    hour_spread: { first: d.utc_first, last: d.utc_last },
  }));

  const varies = new Set(days.map((d) => String(d.actual_ist.first).slice(0, 2))).size > 1;

  const paxBy = new Map(
    (basket as Row[]).map((r) => [`${r.origin}-${r.destination}`, num(r.pax_cy) as number]),
  );
  const paxTotal = [...paxBy.values()].reduce((a, b) => a + b, 0);
  const paxCovered = [...paxBy.entries()]
    .filter(([p]) => withFares.has(p))
    .reduce((a, [, n]) => a + n, 0);
  const nationalRows = (await sql`
    SELECT v FROM basket_meta WHERE k = 'national_total_pax'
  `) as Row[];
  const nationalTotal = num(nationalRows[0]?.v) ?? 0;

  const bf = (backfill as Row[])[0];
  const s = (summary as Row[])[0];
  const hdr = (runHeader as Row[])[0];

  return {
    summary: {
      observations: num(s.observations),
      rows_considered: num(hdr.n_obs_in),
      rows_selected: num(hdr.n_obs_selected),
      collection_runs: num(s.collection_runs),
      sweeps: num(s.sweeps),
      days: days.length,
      sources: (sources as Row[]).map((r) => r.source),
    },
    backfill: {
      rows: num(bf.n),
      date_range: num(bf.n) ? [isoDate(bf.lo), isoDate(bf.hi)] : null,
      sources: (backfillSrc as Row[]).map((r) => r.source),
      routes: (backfillRoutes as Row[]).map((r) => `${r.origin}-${r.destination}`),
      note:
        "Pre-window historical rows imported from other teams' public " +
        'repos/APIs to cover dates before our own collection started ' +
        '(2026-09-10). Not a live sweep, no collection_run behind it, ' +
        'and excluded from the index (config.INDEX_SOURCE keeps the ' +
        'series on one price source). Kept separate from every count ' +
        'and table above so this page still describes only what we ' +
        'actually ran.',
    },
    days,
    clock: {
      nominal_time_ist: NOMINAL_IST,
      schedule:
        'cron 15 5 * * * (05:15 UTC / 10:45 IST), .github/workflows/collect.yml',
      nominal_note:
        'Every day is labelled with the scheduled slot, the ' +
        'way a price is labelled to a reference period rather ' +
        'than to the minute it was written down. The actual ' +
        'clock time is kept beside it and is never overwritten.',
      actual_ist_observed: days.map((d) => d.actual_ist.first),
      max_abs_drift_minutes: days.length
        ? Math.max(...days.map((d) => Math.abs(num(d.drift_minutes) as number)))
        : 0,
      varies_across_days: varies,
      warning: varies
        ? 'Actual collection hour varies across days, so part of the ' +
          'measured movement is the clock, not the market. These were ' +
          'the manual bootstrap days; the schedule above is what ' +
          'unattended runs follow.'
        : null,
    },
    routes: {
      in_basket: inBasket.length,
      attempted: inBasket.filter((r) => attempted.has(r)).length,
      with_fares: inBasket.filter((r) => withFares.has(r)).length,
      never_attempted: inBasket.filter((r) => !attempted.has(r)).sort(),
      attempted_but_no_fares: inBasket
        .filter((r) => attempted.has(r) && !withFares.has(r))
        .sort(),
      basket_pax_covered_pct: Number(((100 * paxCovered) / paxTotal).toFixed(2)),
      national_pax_covered_pct: Number(((100 * paxCovered) / nationalTotal).toFixed(2)),
    },
    sweep_selection: {
      rule:
        'Per (date, route, lead), the latest genuine sweep that covered ' +
        'it wins. Selection is per cell, not per day: 2026-09-10 is the ' +
        'union of four partial sweeps and has no single complete one.',
      excluded_runs: Object.fromEntries(
        (exclusions as Row[]).map((r) => [String(r.run_id), r.reason]),
      ),
      genuineness_screen:
        "A sweep whose fetches arrive faster than the collector's own " +
        'rate limit cannot be live collection. Nothing is excluded at present.',
    },
  };
}

export async function sweeps(indexSource: string, limit = 50) {
  // Span is computed in SQL: a JS Date holds milliseconds, and these
  // timestamps carry microseconds, so subtracting Dates loses precision that
  // shows up in seconds_per_fetch.
  const rows = (await sql`
    SELECT run_id,
           to_char(MIN(started_at) AT TIME ZONE 'UTC', 'YYYY-MM-DD') AS day,
           to_char(MIN(started_at) AT TIME ZONE 'UTC', 'HH24:MI:SS') AS started,
           to_char(MAX(started_at) AT TIME ZONE 'UTC', 'HH24:MI:SS') AS ended,
           EXTRACT(EPOCH FROM (MAX(started_at) - MIN(started_at)))::float8 AS span,
           COUNT(*)::int AS fetches, SUM(n_quotes)::int AS quotes,
           COUNT(*) FILTER (WHERE status = 'OK')::int AS ok,
           COUNT(*) FILTER (WHERE status <> 'OK')::int AS failed,
           MIN(source) AS source
      FROM collection_run GROUP BY run_id ORDER BY MIN(started_at) DESC
     LIMIT ${limit}
  `) as Row[];

  const out = rows.map((r) => {
    const span = num(r.span) as number;
    const fetches = num(r.fetches) as number;
    return {
      run_id: r.run_id,
      date: r.day,
      started: r.started,
      ended: r.ended,
      fetches,
      quotes: num(r.quotes) ?? 0,
      ok: num(r.ok) ?? 0,
      failed: num(r.failed) ?? 0,
      span_seconds: Number(span.toFixed(1)),
      seconds_per_fetch: Number((span / Math.max(fetches - 1, 1)).toFixed(2)),
      source: r.source,
      feeds_index: r.source === indexSource,
      kind: r.source === indexSource ? INDEX_SOURCE_NOTE : 'Base fare / tax study',
    };
  });

  return { n_sweeps: out.length, sweeps: out };
}

export async function runLog(limit = 200) {
  const rows = (await sql`
    SELECT run_id, origin, destination, lead_time_days, departure_date,
           source, status, n_quotes, elapsed_ms, error,
           -- .US keeps the microseconds; a JS Date would truncate to
           -- milliseconds and silently shorten every timestamp.
           to_char(started_at AT TIME ZONE 'UTC', 'YYYY-MM-DD HH24:MI:SS.US')
             AS started_at
      FROM collection_run ORDER BY started_at DESC LIMIT ${limit}
  `) as Row[];
  return {
    n_runs: rows.length,
    runs: rows.map((r) => ({
      run_id: r.run_id,
      route: `${r.origin}-${r.destination}`,
      lead_time_days: num(r.lead_time_days),
      departure_date: isoDate(r.departure_date),
      source: r.source,
      status: r.status,
      n_quotes: num(r.n_quotes),
      elapsed_ms: num(r.elapsed_ms),
      error: r.error,
      started_at: r.started_at,
    })),
  };
}
